### ISSUE TITLE
Shared url maps not always synced

### ISSUE THREAD

**Issue by sa501428** [aidenlab/juicebox.js#280]
Load 2 maps by URL. Example:
https://hicfiles.s3.amazonaws.com/hiseq/gm12878/in-situ/primary.hic
https://hicfiles.s3.amazonaws.com/hiseq/gm12878/in-situ/replicate.hic
Zoom in to some region (maps are synced)

Then generate the shared URL and open it in another tab
Zoom in, and maps are not always synced.
They seem to be synced 30% of the time or so:
http://bit.ly/2wkGh5h

**Comment by jrobinso** [comment]
I played around with your link a bit and couldn't get them un-synched.   Synching can lag while data is being loaded in on or the other maps,  but never more than a second or so for me.    Could you check your java console for errors?

If your computer or network is overloaded or slow I can see that this could be a problem.   Not sure what to do about it though.

**Comment by sa501428** [comment]
I don't think it's a lag though - they completely un-sync :(
Console clear of any errors.
<img width="719" alt="screen shot 2018-08-28 at 12 18 31 pm" src="https://user-images.githubusercontent.com/6503789/44739240-e063a380-aabc-11e8-804c-e6fe5a175000.png">

**Comment by jrobinso** [comment]
So far I can't reproduce it.   If you have some steps to reliably reproduce let me know.

You might also try this link, which is the latest,  the aidenlab instance hasn't been updated in some time.  I doubt you'll see any difference though.

http://bit.ly/2wnN65V

**Comment by sa501428** [comment]
I'm unable to replicate this bug at home, but definitely see it repeatedly at BCM and on different computers... Will try the igv version in lab tomorrow.

**Comment by nchernia** [comment]
This just happened to me as well.

http://bit.ly/2Bhr9qL

**Comment by turner** [comment]
**A candidate mechanism for this, found while scoping candidate 6. Filing it here so it is testable rather than remembered — this is a reading of the code, not a reproduction.**

The report is that two maps loaded by URL are synced, but reopening the shared link syncs them "about 30% of the time." Intermittency across reloads with no user input in between points at a race, and there is one in the load path.

`DataLoader`'s restore ladder has a `synchState` branch that reads, in effect:

```
} else if (config.synchState && this.browser.canBeSynched(config.synchState)) {
    await this.browser.syncState(config.synchState);
    ...
} else {
    state = State.default(config);   // ← unsynced default view
    ...
}
```

**`canBeSynched` is consulted before this browser's dataset is assigned**, and its first substantive check is `if (!this.activeDataset) return false`. So whether a browser honours the `synchState` its own shared URL carries depends on whether some earlier step has already put a dataset on it — and with two maps loading concurrently, that is a race. The browser that loses it takes the `else` branch and opens at `State.default`, unsynced, with **no error and no warning**: the state is valid, just not the one the link asked for.

That would produce exactly the reported shape — mostly-unsynced, occasionally synced, varying run to run on the same link.

## Why it is being written down rather than fixed here

`synchState` is one of the five restore doors that **#557** pins in a golden, so this is cheap to settle as a side effect of work already scheduled rather than as its own investigation. If the hypothesis holds:

- **#559** (`setActiveDataset` loses its `state` parameter) is what forces the ordering to be explicit at every call site — the `synchState` branch is one of the two that today leaves an unvalidated state standing.
- **#562** (the sync trio splits three ways) is where `canBeSynched` moves out to the sync-group module, and where the three separate copies of the `synchable` rule collapse to one. Whatever the ordering fix is, it lands there.

## What would confirm or kill it

Reload a two-map shared link repeatedly and record whether the maps come up synced along with which dataset finished loading first. If synced runs correlate with one particular load order, this is it. If they do not, the mechanism is something else and this comment should be struck rather than left to mislead.

**Not reproduced.** Do not close this issue on the strength of the reading — but do check it when #562 lands, because if it is right, this has been open for eight years and #562 closes it.

**Comment by turner** [comment]
> *This was generated by AI during triage.*

## Triage Notes

Kept open. Labelled `bug` + `needs-triage`. Two corrections to the comment above, both from resolving the links in this thread.

**1. The reproduction links in this thread cannot test the `canBeSynched` hypothesis.**

Both bit.ly chains still resolve, and the payloads survive:

```
?juicebox={hicUrl=…/primary.hic&name=primary.hic&state=3,3,3,0,0,1,NONE&colorScale=575,255,0,0&nvi=…},
          {hicUrl=…/replicate.hic&name=replicate.hic&state=3,3,3,0,0,1,NONE&colorScale=487,255,0,0&nvi=…}
```

**Neither browser block carries a `synchState` field** — not in sa501428's link, not in @nchernia's. These are 2018-vintage shares that predate the field, so the `else`-branch race described above is not the one they would exercise: each browser restores its own explicit `state=`, which in this link is identical for both.

That does not kill the hypothesis. It means the confirmation experiment has to use a **freshly generated two-map share from the current build**, not these links. The check described above — reload repeatedly, record synced-vs-unsynced against which dataset finished loading first — is still the right one; it just needs a link that actually contains `synchState`.

**2. The issues this was parked behind have all closed.**

#557, #559 and #562 are closed. The live sibling is now **#566** — *"The `config.synchState` restore rung is unreachable: `clearDataset()` runs before its guard"* — which is a sharper and already-confirmed statement of the same ordering defect: the guard is consulted at a point where the dataset it depends on is not yet in place. **This issue should be checked when #566 is fixed**, and #566 is where the ordering fix lands.

**3. Unrelated observation on @nchernia's link.**

Its payload begins `hicUrl=+http://hicfiles.s3.amazonaws.com/internal/scale_CTCF_chr1.hic` — a stray `+` (an encoded space) prefixed to the URL. That map would have failed to load outright, so that report may be a different bug from sa501428's rather than a second sighting of the same one.

---

Sibling issues #284, #285, #286 and #299 were closed in the same pass; this is the survivor.

**Comment by turner** [comment]
> *This was generated by AI during triage.*

Moved `needs-triage` → `ready-for-human`.

**Why it can't be delegated:** the next step is an experiment a person has to run and interpret — generate a fresh two-map share from the current build, reload it repeatedly, and record synced-vs-unsynced against which dataset finished loading first. The 2018 links in this thread cannot serve, for the reason given above: they carry no `synchState`.

Until that runs, the mechanism is a reading of the code rather than a diagnosis, and there is nothing for an agent to implement. If the hypothesis holds, the fix lands in #566.

**Comment by turner** [comment]
## The `synchState` hypothesis is retired

This issue has been pointed at the `config.synchState` rung of `dataLoader.loadHicFile`, on the reading that it consults its guard before the dataset is assigned and so loses a load race. Two comments here say the fix lands in #566.

**It does not, and the rung is gone.** #566 established the rung was not racy but *unreachable* — `loadHicFile` opens with `clearDataset()`, four lines above a guard that returns false without a dataset, so every config carrying a `synchState` took the fallback always, not 30% of the time. A rung nothing takes cannot produce intermittency.

#566 then deleted it rather than repairing it: it arrived in `830f4e7` (Sep 2017, "Set initial synch state of new browser panels") and was superseded three months later by `b71d163` ("sync on map load, not browser creation"), and nothing has supplied the key since. So there is no longer a #566 fix to wait for.

## Where the hypothesis moves

The order-dependent code on a two-map load is the step that replaced the rung, at the end of `loadHicFile`:

```js
registry.sync();

const peer = registry.browsers.find(
    b => b !== this.browser &&
         b.dataset &&
         b.dataset.isCompatible(this.browser.dataset)
);
if (peer) {
    await this.browser.syncState(peer.getSyncState());
}
```

Whichever browser finishes loading first finds no compatible peer and opens at its own state; the second finds the first and syncs to it. That is genuinely a function of which load finishes first, which is what "sometimes" would have to come from.

**This is still a reading, not a diagnosis.** What this issue needs is the repro — two maps by URL, shared link, reopened — run against current `master`, not a third hypothesis. Staying open and `ready-for-human` for that reason.

**Comment by turner** [comment]
## Closing — not reproducible on current `master`

The experiment this issue was held open for has been run: two maps loaded by URL on current `master` (`78dcc20`, v4.2.0), share link generated from that build, reopened repeatedly. The maps came up synced. No de-sync observed.

That is the evidence the last three comments here were waiting on, so the issue closes on a result rather than on age.

**On the outstanding hypothesis.** The order-dependent code at the end of `dataLoader.loadHicFile` — first browser to finish loading finds no compatible peer and keeps its own state, second finds the first and syncs to it — is left as written. It was never observed producing a de-sync; it was a reading of the code offered as a candidate mechanism, and the repro that would have confirmed it did not. It is not being called fixed, because nothing was fixed. It is being called *not a live defect*, on eight years without a sighting and a clean run against the current build.

If a de-sync on a shared two-map link is ever seen again, that peer-selection step is where to look first, and this issue is the history.
